contracts/src/VectorX.sol

```diff
@@ -78,35 +78,81 @@ contract VectorX is IVectorX, TimelockedUpgradeable {
         frozen = _freeze;
     }
 
-    /// @notice TODO: This should be removed for the mainnet release.
+    /// @notice Update the function IDs.
+    function updateFunctionIds(
+        bytes32 _headerRangeFunctionId,
+        bytes32 _rotateFunctionId
+    ) external onlyGuardian {
+        headerRangeFunctionId = _headerRangeFunctionId;
+        rotateFunctionId = _rotateFunctionId;
+    }
+
+    /// @notice Update the gateway address.
+    function updateGateway(address _gateway) external onlyGuardian {
+        gateway = _gateway;
+    }
+
+    /// @notice Update the genesis state of the light client.
     function updateGenesisState(
-        address _gateway,
         uint32 _height,
         bytes32 _header,
         uint64 _authoritySetId,
-        bytes32 _authoritySetHash,
-        bytes32 _headerRangeFunctionId,
-        bytes32 _rotateFunctionId
+        bytes32 _authoritySetHash
     ) external onlyGuardian {
-        gateway = _gateway;
         blockHeightToHeaderHash[_height] = _header;
         authoritySetIdToHash[_authoritySetId] = _authoritySetHash;
         latestBlock = _height;
-        headerRangeFunctionId = _headerRangeFunctionId;
-        rotateFunctionId = _rotateFunctionId;
     }
 
-    /// @notice Request a header update and data commitment from range (trustedBlock, requestedBlock].
-    /// @param _trustedBlock The block height of the trusted block.
-    /// @param _authoritySetId The authority set id of the header range (trustedBlock, requestedBlock].
+    /// @notice Force update the data & state commitments for a range of blocks.
+    function updateBlockRangeData(
+        uint32[] calldata _startBlocks,
+        uint32[] calldata _endBlocks,
+        bytes32[] calldata _headerHashes,
+        bytes32[] calldata _dataRootCommitments,
+        bytes32[] calldata _stateRootCommitments,
+        uint64 _endAuthoritySetId,
+        bytes32 _endAuthoritySetHash
+    ) external onlyGuardian {
+        assert(
+            _startBlocks.length == _endBlocks.length &&
+                _endBlocks.length == _headerHashes.length &&
+                _headerHashes.length == _dataRootCommitments.length &&
+                _dataRootCommitments.length == _stateRootCommitments.length
+        );
+
+        for (uint256 i = 0; i < _startBlocks.length; i++) {
+            assert(_startBlocks[i] == latestBlock);
+            bytes32 key = keccak256(
+                abi.encode(_startBlocks[i], _endBlocks[i])
+            );
+            dataRootCommitments[key] = _dataRootCommitments[i];
+            stateRootCommitments[key] = _stateRootCommitments[i];
+
+            blockHeightToHeaderHash[_endBlocks[i]] = _headerHashes[i];
+            latestBlock = _endBlocks[i];
```

Comment:
why don't we move this out of the form loop? causes an unecesary sstore every time?

Reply:
Removing this, the guardian/multi-sig will need to make sure that it doesn't submit bad data.

Reply:
Found a better way to do it.